src/azul/openapi.py

```diff
@@ -0,0 +1,113 @@
+import copy
+from typing import MutableMapping, List
+
+from azul.deployment import aws
+from azul.types import JSON
+
+
+def spec(description: JSON):
+    """
+    Add description to decorated function
+    >>> @spec({'a': 'b'})
+    ... def foo():
+    ...     pass
+    ...
+
+    >>> foo.api_spec == {'a': 'b'}
+    True
+    """
+
+    def spec_adder(func):
+        func.api_spec = description
+        return func
+    return spec_adder
+
+
+def annotated_specs(gateway_id, app, api_info) -> JSON:
+    """
+    Finds all routes in app that are decorated with @spec and adds this information
+    into the api spec downloaded from API Gateway.
+    :param gateway_id: API Gateway ID corresponding to the Chalice app
+    :param app: App with annotated routes
+    :param api_info: Top level OpenAPI info, definitions, etc.
+    :return: The annotated specifications
+    """
+    specs = aws.api_gateway_export(gateway_id)
+    clean_specs(specs)
+    # TODO: if we want to set version, override needs to be true.
+    specs = merge_dicts(specs, api_info)
+    docs = get_doc_specs(app)
+    for path, doc_spec in docs.items():
+        for verb, description in specs['paths'][path].items():
+            description = merge_dicts(description, doc_spec)
+            specs['paths'][path][verb] = description
+    return specs
+
+
+def clean_specs(specs):
+    """
+    Adjust specs from API Gateway so they pass linting
+    """
+    # Filter out 'options' since it causes linting errors
+    for path in specs['paths']:
+        specs['paths'][path].pop('options', None)
+
+
+def get_doc_specs(app):
+    docs = {}
+    for route, entries in app.routes.items():
+        func = next(iter(entries.values())).view_function
+        if hasattr(func, 'api_spec'):
+            docs[route] = func.api_spec
+    return docs
+
+
+def merge_dicts(d1: MutableMapping, d2: MutableMapping, mutate: bool = False, override: bool = False):
+    """
+    Merge two dictionaries deeply such that:
+    - collisions raise an error,
+    - lists are appended, and
+    - child dictionaries are recursively merged
+    If override is true, d2 overrides d1 on collisions.
+    >>> merge_dicts({1: 'a'}, {2: 'a'})
+    {1: 'a', 2: 'a'}
+    >>> merge_dicts({1: 1}, {1: 2})
+    Traceback (most recent call last):
+    ...
+    ValueError: Cannot merge path '1' with values: 1, 2
+    >>> merge_dicts({1: 1}, {1: 2}, override=True)
+    {1: 2}
+    >>> merge_dicts({1: [1, 2]}, {1: [2, 3]})
+    {1: [1, 2, 2, 3]}
+    >>> merge_dicts({1: {'a': 'z', 'c': [1]}}, {1: {'b': 'y', 'c': [2]}})
+    {1: {'a': 'z', 'c': [1, 2], 'b': 'y'}}
+    >>> merge_dicts({1: {'a': 'z'}}, {1: {'a': 'y'}})
+    Traceback (most recent call last):
+    ...
+    ValueError: Cannot merge path '1.a' with values: z, y
+    >>> merge_dicts({1: {'a': 'z'}}, {1: {'a': 'y'}}, override=True)
+    {1: {'a': 'y'}}
+    """
+    if mutate:
```

Comment:
I don't see this ever being used. If not, just eliminate it. If it is, the naming is odd. mutate=True cause a copy to be made? I'd expect the opposite, mutate=False should make the copy. Also if the function modifies the argument, it should not return anything. So: eliminate if not used or replace this modal argument with a make a wrapper function that copies first.

Reply:
BTW, `inplace` is a more common name for this behavior.

Reply:
Woops! The opposite behavior was what was intended. I'll get rid of it then... I don't believe it's covered in doctests. I made it originally because I wasn't sure if I'd need it or not.